File: 백트래킹/BOJ_15656.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n')
let [n, m] = input[0].split(' ').map(Number)
let arr = input[1].split(' ').map(Number).sort((a, b) => a - b);

var res = [];
var answer = [];
const dfs = (depth) => {
    if(depth === m){
        answer.push(res.join(' ') + '\n');
        return
    }
    
    for(num of arr){
        res.push(num);
        dfs(depth + 1);
        res.pop();
    }
}

dfs(0);
console.log(answer.join(''));

File: 백트래킹/BOJ_15656.py
# n과m (7), S3
from sys import stdin

n, m = map(int, stdin.readline().split())
arr = sorted(list(map(int, stdin.readline().split())))

res = []
def dfs(depth):
    if depth == m:
        print(*res)
        return

    for i in range(n):
        res.append(arr[i])
        dfs(depth + 1)
        res.pop()

dfs(0)